### ISSUE TITLE
Multiple webhook callbacks return duplicate trace ids when multiple requests are sent to Gotenberg at the same time

### ISSUE THREAD

**Issue by lrakauskas** [gotenberg/gotenberg#1000]
## General Description
Webhook callbacks tend to return duplicate trace ids when there are multiple requests pending on Gotenberg. We've experienced this on `/forms/libreoffice/convert` endpoint in production, but when testing it, it happens on `/forms/chromium/convert/url`, which makes me believe that's the general issue with webhooks. See repro steps for better understanding.

## Repro Steps
1. `docker network create sample-app` - create shared network container
2. `docker run --rm -p 3000:3000 --network sample-app gotenberg/gotenberg:8` - run Gotenberg
3. `docker run -p 8080:8080 --rm -t --network sample-app --name http-logger mendhak/http-https-echo` - run any container of your liking to log all http requests
4. (optional) `docker logs -f http-logger | grep "gotenberg-trace"` - filter for `gotenberg-trace` header for the ease of reading. You can skip this if you want to read docker logs manually.
5. `for i in $(seq 1 10); do curl --request POST http://localhost:3000/forms/chromium/convert/url --header "Gotenberg-Trace: $i" --header 'Gotenberg-Webhook-Url: http://http-logger:8080' --header 'Gotenberg-Webhook-Error-Url: http://http-logger:8080' --header 'Gotenberg-Webhook-Method: POST' --form url=https://sparksuite.github.io/simple-html-invoice-template/ & done` - send 10 parallel requests to Gotenberg to request URL to PDF conversions, with webhook urls pointing to our container from step 3. We'll use iteration number `$i` as Gotenberg-Trace value. "10" is an arbitrary number, sometimes you'll need to try this multiple times, or you can bump the for loop to a 100.

### Results
Headers log from a container hosting a webhook:
![image](https://github.com/user-attachments/assets/c095cc1f-289b-487f-bdc1-77fd4ef91a40)
Webhook was called 10 times as expected, but trace values should've been 1 to 10 in any order. It tends to happen quite randomly. Sometimes there's only two values clashing, sometimes it works.

If we look at logs from Gotenberg - everything looks correct here and it thinks that it called it with correct headers.
```
{"level":"info","ts":1728066147.451547,"logger":"api","msg":"request handled","trace":"2","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":281866,"latency_human":"281.866µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4518194,"logger":"api","msg":"request handled","trace":"4","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":223847,"latency_human":"223.847µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4519765,"logger":"api","msg":"request handled","trace":"6","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":353824,"latency_human":"353.824µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4520564,"logger":"api","msg":"request handled","trace":"1","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":344227,"latency_human":"344.227µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4523084,"logger":"api","msg":"request handled","trace":"7","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":193156,"latency_human":"193.156µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.452975,"logger":"api","msg":"request handled","trace":"9","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":157636,"latency_human":"157.636µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4534066,"logger":"api","msg":"request handled","trace":"3","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":220205,"latency_human":"220.205µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.453535,"logger":"api","msg":"request handled","trace":"5","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":236851,"latency_human":"236.851µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.453616,"logger":"api","msg":"request handled","trace":"8","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":122463,"latency_human":"122.463µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.4542732,"logger":"api","msg":"request handled","trace":"10","remote_ip":"172.19.0.1","host":"localhost:3000","uri":"/forms/chromium/convert/url","method":"POST","path":"/forms/chromium/convert/url","referer":"","user_agent":"curl/7.81.0","status":204,"latency":148272,"latency_human":"148.272µs","bytes_in":196,"bytes_out":0}
{"level":"info","ts":1728066147.5548465,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"2","webhook_url":"http://http-logger:8080","method":"POST","latency":103584101,"latency_human":"103.584101ms","bytes_out":111721}
{"level":"info","ts":1728066147.6532495,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"4","webhook_url":"http://http-logger:8080","method":"POST","latency":201654715,"latency_human":"201.654715ms","bytes_out":111721}
{"level":"info","ts":1728066147.7542367,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"6","webhook_url":"http://http-logger:8080","method":"POST","latency":302614171,"latency_human":"302.614171ms","bytes_out":111721}
{"level":"info","ts":1728066147.851595,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"1","webhook_url":"http://http-logger:8080","method":"POST","latency":399882042,"latency_human":"399.882042ms","bytes_out":111721}
{"level":"info","ts":1728066147.952587,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"7","webhook_url":"http://http-logger:8080","method":"POST","latency":500472421,"latency_human":"500.472421ms","bytes_out":111721}
{"level":"info","ts":1728066148.054777,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"9","webhook_url":"http://http-logger:8080","method":"POST","latency":601960257,"latency_human":"601.960257ms","bytes_out":111721}
{"level":"info","ts":1728066148.157313,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"3","webhook_url":"http://http-logger:8080","method":"POST","latency":704126954,"latency_human":"704.126954ms","bytes_out":111721}
{"level":"info","ts":1728066148.2572732,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"5","webhook_url":"http://http-logger:8080","method":"POST","latency":803974974,"latency_human":"803.974974ms","bytes_out":111721}
{"level":"info","ts":1728066148.3550417,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"8","webhook_url":"http://http-logger:8080","method":"POST","latency":901548711,"latency_human":"901.548711ms","bytes_out":111721}
{"level":"info","ts":1728066148.4535835,"logger":"api.formschromiumconverturl","msg":"request to webhook handled","trace":"10","webhook_url":"http://http-logger:8080","method":"POST","latency":999458672,"latency_human":"999.458672ms","bytes_out":111721}
```

## Final comments
I believe there's a race condition or some sort of static variable somewhere in Gotenberg? Note that we've ruled out webhook endpoint being an issue because in production we use Laravel and here in repro steps I've just picked random minimalistic docker image that only logs requests.

To get around this in our current production server - we've stopped using webhooks at all and we're just doing sync requests.

P.S. really lovely project you have here, big fans.

**Comment by gulien** [comment]
Hey @lrakauskas 👋 

Thanks for the kind words!

Regarding your issue, it looks like I’m using a context which is being recycled by the server library I’m using 🤔 the fix should be easy though!

**Comment by gulien** [comment]
@lrakauskas Image `gotenberg/snapshot:fixwebhook-recycled-trace` is available with a fix 😉 Let me know if it's working as expected 😄

**Comment by lrakauskas** [comment]
@gulien Yup, can confirm, works flawlessly now. Thank you! Great response time 😄

**Comment by gulien** [comment]
Thanks for the quick feedback @lrakauskas 🎉 

`gotenberg/gotenberg:edge` should be available within ~1 hour.

**Comment by lrakauskas** [comment]
@gulien  Is this going to be merged into 8 at some point?

**Comment by gulien** [comment]
I may release a new version tomorrow.

**Comment by gulien** [comment]
@lrakauskas version `8.11.1` should [soon](https://github.com/gotenberg/gotenberg/actions/runs/11200803326) be available.
